README.md
[![Build Status](https://travis-ci.com/6eDesign/timeUtils.svg?branch=master)](https://travis-ci.com/6eDesign/timeUtils)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)
[![npm version](https://badge.fury.io/js/timeUtils.svg)](https://badge.fury.io/js/timeUtils)

# timeUtils.js
Simple time/date formatter for JavaScript which uses PHP-like syntax to easily format JavaScript date objects into legible/formatted strings.  It works in the browser and Node.js and is quite small ~1.3KB (minified & gzipped). 

## Installation & Including the Script:

First, install the module via NPM: 

```npm i timeUtils```[<sup>1</sup>](#1)

Then include the module...

As an import: 

```javascript
import { formatDate } from 'timeUtils'
```

As a Common JS library: 
```javascript
const formatDate = require('timeUtils').formatDate; 
```

As a script tag: 
```html
<script type="text/javascript" src="node_modules/timeUtils/dist/timeUtils.umd.js" />
```

## Usage:

timeUtils's `formatDate` function expects a JavaScript Date object as its first argument and a string-based template as its second argument.  

The templates should take the following form: 
`"Lorem Ipsum '#{m}/#{d}/#{Y}'"`

Where strings wrapped in `#{}` represent date/time variables which  will be formatted for you.  The  table below shows all the variables currently available along with a brief description.  In the above example template, we would get date in the "mm/dd/yyyy" format - for example.

### Date Variables
| Symbol/Token  | Type | Description | Example(s) |
| ------------- | ------------- | ------------- | ------------- |
| D | day of week | Textual representation of day of week, 3 letters | Sun, Mon, ... Sat |
| l | day of week | Full textual representation of day of week | Sunday, Monday, ... Saturday |
| F | month | Full textual representation of the month | January, ... December |
| M | month | Abbreviated textual representation of month, 3 letters | Jan, Feb, ... Dec |
| n | month | Numeric representation of month without leading 0's | 1, 2, ... 12 |
| m | month | Numeric representation of the month with leading 0's | 01, 02, ... 12 |
| j | day of month | Day of month, without leading 0's | 1, 2, ... 31 |
| d | day of month | Day of the month, 2 digits, with leading zeros | 01, 02, ... 31 |
| y | year | Short numeric year, 2 digits | 00, 01, ... 15 |
| Y | year | Full numeric year, 4 digits | 2000, 2001, ... 2015 |

### Time Variables
Symbol/Token  | Type | Description | Example(s)
------------- | ------------- | ------------- | -------------
| g | hour | 12-hour format of the hour, without leading 0's | 1, 2, ... 12 |
| h | hour | 12-hour format of the hour, with leading 0's | 01, 02, ... 12 |
| G | hour | 24-hour format of the hour, without leading 0's | 0, 1, ... 23 |
| H | hour | 24-hour format of the hour, with leading 0's | 00, 01, ... 23 |
| i | minute | Minutes with leading 0's | 00, 01, ... 59 |
| s | seconds | Seconds with leading 0's | 00, 01, ... 59 |
| a | am/pm | Lowercase ante meridiem and post meridiem | 'am' or 'pm' |
| A | am/pm | Uppercase ante meridiem and post meridiem | 'AM' or 'PM' |

### Internationalization: 
timeUtils provides an `internationalize` method which can be used to support multiple languages: 

```javascript
import { internationalize } from 'timeUtils';
internationalize({
  daysOfWeek: [
    'Domingo',
    'Lunes',
    'Martes',
    'Miércoles',
    'Jueves',
    'Viernes',
    'Sábado'
  ],
  monthsOfYear: [
    'Enero',
    'Febrero',
    'Marzo',
    'Abril',
    'Mayo',
    'Junio',
    'Julio',
    'Agosto',
    'Septiembre',
    'Octubre',
    'Noviembre',
    'Diciembre'
  ]
})
```

### Examples
```javascript 
var date = new Date()
```

> Mon Apr 06 2015 21:51:08 GMT-0600 (Mountain Daylight Time)

```javascript
timeUtils.formatDate(date,'Your appointment is on #{l}, #{F} #{j}, #{Y} at #{g}:#{i} #{A}.')
```

> "Your appointment is on Monday, April 6, 2015 at 9:48 PM."

<a class="anchor" id="1"></a><sup>1</sup> add --save if you are using npm < 5.0.0


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.1.6"></a>
## [1.1.6](https://github.com/6eDesign/timeUtils/compare/v1.1.5...v1.1.6) (2019-11-15)



<a name="1.1.5"></a>
## [1.1.5](https://github.com/6eDesign/timeUtils/compare/v1.1.4...v1.1.5) (2019-07-15)


### Bug Fixes

* set package.browser to ESM build ([98f5f78](https://github.com/6eDesign/timeUtils/commit/98f5f78))



<a name="1.1.4"></a>
## [1.1.4](https://github.com/6eDesign/timeUtils/compare/v1.1.3...v1.1.4) (2018-12-04)


### Bug Fixes

* add uglify to rollup config ([8837706](https://github.com/6eDesign/timeUtils/commit/8837706))



<a name="1.1.3"></a>
## [1.1.3](https://github.com/6eDesign/timeUtils/compare/v1.1.2...v1.1.3) (2018-12-04)


### Bug Fixes

* Add internationalization documentation and tests ([1e6437c](https://github.com/6eDesign/timeUtils/commit/1e6437c))
* test formatting ([3acd757](https://github.com/6eDesign/timeUtils/commit/3acd757))
* Update release script ([16983bd](https://github.com/6eDesign/timeUtils/commit/16983bd))



<a name="1.1.2"></a>
## [1.1.2](https://github.com/6eDesign/timeUtils/compare/v1.1.1...v1.1.2) (2018-12-03)


### Bug Fixes

* minify dist/ files ([d3979a7](https://github.com/6eDesign/timeUtils/commit/d3979a7))
* minify dist/ files ([c9a94bc](https://github.com/6eDesign/timeUtils/commit/c9a94bc))
* minify dist/ files ([abe1252](https://github.com/6eDesign/timeUtils/commit/abe1252))
* update release script ([0c818a9](https://github.com/6eDesign/timeUtils/commit/0c818a9))



<a name="1.1.1"></a>
## [1.1.1](https://github.com/6eDesign/timeUtils/compare/v1.1.0...v1.1.1) (2018-12-03)


### Bug Fixes

* Updated documentation ([edcca1a](https://github.com/6eDesign/timeUtils/commit/edcca1a))



<a name="1.1.0"></a>
# 1.1.0 (2018-12-03)


### Features

* convert to rollup ([81580f3](https://github.com/6eDesign/timeUtils/commit/81580f3))


test/index.js
var test = require('tape'); 
var timeUtils = require('../dist/timeUtils.cjs.js');
var formatterAssertions = require('./data/formatterAssertions');

test('Date formatter tests: ', (t) => {
  t.plan(formatterAssertions.length); 
  formatterAssertions.forEach(assertion => { 
    let val = timeUtils.formatDate(assertion.date,assertion.format); 
    t.equal(val,assertion.formatted,`Expected ${val} to equal ${assertion.formatted}`);
  });
});

tests/internationalize.js
import test from 'tape';
import { internationalize, formatDate } from '../dist/timeUtils.esm.js'; 
import formatterAssertions from './lib/formatterAssertions';

const daysOfWeek = [
  'Domingo',
  'Lunes',
  'Martes',
  'Miércoles',
  'Jueves',
  'Viernes',
  'Sábado'
];
const monthsOfYear = [
  'Enero',
  'Febrero',
  'Marzo',
  'Abril',
  'Mayo',
  'Junio',
  'Julio',
  'Agosto',
  'Septiembre',
  'Octubre',
  'Noviembre',
  'Diciembre'
];


test('Date formatter w/ internationalization tests: ', (t) => {
  t.plan(formatterAssertions.length); 
  internationalize({
    daysOfWeek,
    monthsOfYear
  });
  formatterAssertions.forEach(assertion => { 
    let val = formatDate(assertion.date,assertion.format); 
    t.equal(val,assertion.formattedSpanish,`Expected ${val} to equal ${assertion.formattedSpanish}`);
  });
});

test/data/formatterAssertions.js
let date1 = new Date(2018,8,29,9,30,0,0);
let date2 = new Date(2018,11,03,16,30,0,0);
let date3 = new Date(2018,11,03,0,0,0,0);

module.exports = [ 
  { 
    date: date1,
    format: '#{D} #{M} #{d}, #{Y}', 
    formatted: 'Sat Sep 29, 2018'
  }, { 
    date: date1,
    format: '#{l} #{F} #{d}, #{y}', 
    formatted: 'Saturday September 29, 18'
  }, { 
    date: date2, 
    format: '#{l} #{F} #{d}, #{Y}', 
    formatted: 'Monday December 03, 2018'
  }, { 
    date: date2, 
    format: '#{l} #{F} #{j}, #{Y}',
    formatted: 'Monday December 3, 2018'
  }, { 
    date: date1, 
    format: '#{m}/#{d}/#{Y}',
    formatted: '09/29/2018'
  }, { 
    date: date1, 
    format: '#{n}/#{d}/#{Y}',
    formatted: '9/29/2018'
  }, { 
    date: date1, 
    format: '#{g}:#{i}:#{s} #{a}',
    formatted: '9:30:00 am'
  }, { 
    date: date1, 
    format: '#{G}:#{i}:#{s} #{a}',
    formatted: '9:30:00 am'
  }, { 
    date: date1, 
    format: '#{h}:#{i}:#{s} #{a}',
    formatted: '09:30:00 am'
  }, { 
    date: date1, 
    format: '#{H}:#{i}:#{s} #{a}',
    formatted: '09:30:00 am'
  }, { 
    date: date2, 
    format: '#{g}:#{i}:#{s} #{a}',
    formatted: '4:30:00 pm'
  }, { 
    date: date2, 
    format: '#{G}:#{i}:#{s} #{a}',
    formatted: '16:30:00 pm'
  }, { 
    date: date3, 
    format: '#{G}:#{i}:#{s} #{a}',
    formatted: '0:00:00 am'
  }, { 
    date: date3, 
    format: '#{H}:#{i}:#{s} #{a}',
    formatted: '00:00:00 am'
  }
];

gulpfile.js
// gulp task dependencies
var gulp = require('gulp');
var uglify = require('gulp-uglify');
var concat = require('gulp-concat');
var less = require('gulp-less');
var minifyCSS = require('gulp-minify-css');
var livereload = require('gulp-livereload');
var path = require('path');

// 
// BASE PREFIX
//

var input = './';
var output = './';


// 
// DIST PATHS
// 

var dist = {
  outfile: 'timeUtils.min.js',
  files: [
    input + 'timeUtils.js'
  ]
};

//
// DIST TASK
//

gulp.task('scripts:dist', function() {
  return gulp.src(dist.files)
    .pipe(uglify())
    .pipe(concat(dist.outfile))
    .pipe(gulp.dest(output))
    .pipe(livereload());
});



//
// WATCHER TASK
//

gulp.task('watch', function() {
  gulp.watch(dist.files, ['scripts:dist']);
});

//
// TASK ALIAS'
//

gulp.task('default', ['scripts:dist']);
gulp.task('uber', ['scripts:dist', 'watch']);


src/index.js
import { injectStringData, enforceLength } from './lib/utils.js';
import { dictionary, extendDictionary, resetDictionary } from './config.js';

var acceptedDateTokens = [
  { 
    // d: day of the month, 2 digits with leading zeros:
    key: 'd', 
    method: function(date) { return enforceLength(date.getDate(), 2); } 
  }, { 
    // D: textual representation of day, 3 letters: Sun thru Sat
    key: 'D', 
    method: function(date) { return enforceLength(dictionary.daysOfWeek[date.getDay()],3); } 
  }, { 
    // j: day of month without leading 0's
    key: 'j', 
    method: function(date) { return date.getDate(); } 
  }, { 
    // l: full textual representation of day of week: Sunday thru Saturday
    key: 'l', 
    method: function(date) { return dictionary.daysOfWeek[date.getDay()]; } 
  }, { 
    // F: full text month: 'January' thru 'December'
    key: 'F', 
    method: function(date) { return dictionary.monthsOfYear[date.getMonth()]; } 
  }, { 
    // m: 2 digit numeric month: '01' - '12':
    key: 'm', 
    method: function(date) { return enforceLength(date.getMonth()+1,2); } 
  }, { 
    // M: a short textual representation of the month, 3 letters: 'Jan' - 'Dec'
    key: 'M', 
    method: function(date) { return enforceLength(dictionary.monthsOfYear[date.getMonth()],3); } 
  }, { 
    // n: numeric represetation of month w/o leading 0's, '1' - '12':
    key: 'n', 
    method: function(date) { return date.getMonth() + 1; } 
  }, { 
    // Y: Full numeric year, 4 digits
    key: 'Y', 
    method: function(date) { return date.getFullYear(); } 
  }, { 
    // y: 2 digit numeric year:
    key: 'y', 
    method: function(date) { return enforceLength(date.getFullYear(),2,true); }
   }
];

var acceptedTimeTokens = [
  { 
    // a: lowercase ante meridiem and post meridiem 'am' or 'pm'
    key: 'a', 
    method: function(date) { return (date.getHours() > 11) ? 'pm' : 'am'; } 
  }, { 
    // A: uppercase ante merdiiem and post meridiem 'AM' or 'PM'
    key: 'A', 
    method: function(date) { return (date.getHours() > 11) ? 'PM' : 'AM'; } 
  }, { 
    // g: 12-hour format of an hour without leading zeros 1-12
    key: 'g', 
    method: function(date) { return date.getHours() % 12 || 12; } 
  }, { 
    // G: 24-hour format of an hour without leading zeros 0-23
    key: 'G', 
    method: function(date) { return date.getHours(); } 
  }, { 
    // h: 12-hour format of an hour with leading zeros 01-12
    key: 'h', 
    method: function(date) { return enforceLength(date.getHours()%12 || 12,2); } 
  }, { 
    // H: 24-hour format of an hour with leading zeros: 00-23
    key: 'H', 
    method: function(date) { return enforceLength(date.getHours(),2); } 
  }, { 
    // i: Minutes with leading zeros 00-59
    key: 'i', 
    method: function(date) { return enforceLength(date.getMinutes(),2); } 
  }, { 
    // s: Seconds with leading zeros 00-59
    key: 's', 
    method: function(date) { return enforceLength(date.getSeconds(),2); }
   }
];

/**
 * Internationalization object for timeUtils.internationalize().
 * @typedef internationalizeObj
 * @property {Array} [daysOfWeek=[ 'Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday' ]] daysOfWeek Weekday labels as strings, starting with Sunday.
 * @property {Array} [monthsOfYear=[ 'January','February','March','April','May','June','July','August','September','October','November','December' ]] monthsOfYear Month labels as strings, starting with January.
 */

/**
 * This function can be used to support additional languages by passing an object with 
 * `daysOfWeek` and `monthsOfYear` attributes.  Each attribute should be an array of
 * strings (ex: `daysOfWeek: ['monday', 'tuesday', 'wednesday'...]`)
 *
 * @param {internationalizeObj} conf
 */
export const internationalize = (conf={}) => { 
  extendDictionary(conf);
};

/**
 * generic formatDate function which accepts dynamic templates
 * @param date {Date} Required
 * @param template {String} Optional
 * @returns {String}
 *
 * @example
 * formatDate(new Date(), '#{M}. #{j}, #{Y}')
 * @returns {Number} Returns a formatted date
 *
 */
export const formatDate = (date,template='#{m}/#{d}/#{Y}') => {
  acceptedDateTokens.forEach(token => {
    if(template.indexOf(`#{${token.key}}`) == -1) return; 
    template = injectStringData(template,token.key,token.method(date))
  }); 
  acceptedTimeTokens.forEach(token => {
    if(template.indexOf(`#{${token.key}}`) == -1) return;
    template = injectStringData(template,token.key,token.method(date));
  });
  return template;
};

/**
 * Small function for resetting language to English (used in testing).
 */
export const resetInternationalization = () => resetDictionary();

timeUtils.js
var timeUtils = (function(pub){

  /**
   * generic function to inject data into token-laden string
   * @param str {String} Required
   * @param name {String} Required
   * @param value {String|Integer} Required
   * @returns {String}
   *
   * Pass this function a string with encoded tokens, example: 
   *    util.injectStringData("The following is a token: #{tokenName}", "tokenName", 123); 
   *   
   *    Returns: "The following is a token: 123"
   *
   */
  pub.injectStringData = function(str,name,value) {
    return str.replace(new RegExp('#{'+name+'}','g'),value);
  };

  /**
   * generic function to enforce length of string
   * @param str {String} Required
   * @param length {Integer} Required
   * @param fromBack {Boolean} Optional
   * @returns {String}
   *
   * Pass a string or number to this function and specify the desired length.
   * This function will either pad the # with leading 0's (if str.length < length)
   * or remove data from the end (@fromBack==false) or beginning (@fromBack==true)
   * of the string when str.length > length.
   *
   * When length == str.length or typeof length == 'undefined', this function
   * returns the original @str parameter.
   *
   */
  var enforceLength = function(str,length,fromBack) {
    str = str.toString();
    if(typeof length == 'undefined') return str;
    if(str.length == length) return str;
    fromBack = (typeof fromBack == 'undefined') ? false : fromBack;
    if(str.length < length) {
      // pad the beginning of the string w/ enough 0's to reach desired length:
      while(length - str.length > 0) str = '0' + str;
    } else if(str.length > length) {
      if(fromBack) {
        // grab the desired #/chars from end of string: ex: '2015' -> '15'
        str = str.substring(str.length-length);
      } else {
        // grab the desired #/chars from beginning of string: ex: '2015' -> '20'
        str = str.substring(0,length);
      }
    }
    return str;
  };

  // Internal variables for storing days of week, months of year: 
  var daysOfWeek = [ 'Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday' ];
  var monthsOfYear = [ 'January','February','March','April','May','June','July','August','September','October','November','December'];


  var acceptedDateTokens = [
    // d: day of the month, 2 digits with leading zeros:
    { key: 'd', method: function(date) { return enforceLength(date.getDate(), 2); } },
    // D: textual representation of day, 3 letters: Sun thru Sat
    { key: 'D', method: function(date) { return enforceLength(daysOfWeek[date.getDay()],3); } },
    // j: day of month without leading 0's
    { key: 'j', method: function(date) { return date.getDate(); } },
    // l: full textual representation of day of week: Sunday thru Saturday
    { key: 'l', method: function(date) { return daysOfWeek[date.getDay()]; } },
    // F: full text month: 'January' thru 'December'
    { key: 'F', method: function(date) { return monthsOfYear[date.getMonth()]; } },
    // m: 2 digit numeric month: '01' - '12':
    { key: 'm', method: function(date) { return enforceLength(date.getMonth()+1,2); } },
    // M: a short textual representation of the month, 3 letters: 'Jan' - 'Dec'
    { key: 'M', method: function(date) { return enforceLength(monthsOfYear[date.getMonth()],3); } },
    // n: numeric represetation of month w/o leading 0's, '1' - '12':
    { key: 'n', method: function(date) { return date.getMonth() + 1; } },
    // Y: Full numeric year, 4 digits
    { key: 'Y', method: function(date) { return date.getFullYear(); } },
    // y: 2 digit numeric year:
    { key: 'y', method: function(date) { return enforceLength(date.getFullYear(),2,true); } }
  ];

  var acceptedTimeTokens = [
    // a: lowercase ante meridiem and post meridiem 'am' or 'pm'
    { key: 'a', method: function(date) { return (date.getHours() > 11) ? 'pm' : 'am'; } },
    // A: uppercase ante merdiiem and post meridiem 'AM' or 'PM'
    { key: 'A', method: function(date) { return (date.getHours() > 11) ? 'PM' : 'AM'; } },
    // g: 12-hour format of an hour without leading zeros 1-12
    { key: 'g', method: function(date) { return date.getHours() % 12 || 12; } },
    // G: 24-hour format of an hour without leading zeros 0-23
    { key: 'G', method: function(date) { return date.getHours(); } },
    // h: 12-hour format of an hour with leading zeros 01-12
    { key: 'h', method: function(date) { return enforceLength(date.getHours()%12 || 12,2); } },
    // H: 24-hour format of an hour with leading zeros: 00-23
    { key: 'H', method: function(date) { return enforceLength(date.getHours(),2); } },
    // i: Minutes with leading zeros 00-59
    { key: 'i', method: function(date) { return enforceLength(date.getMinutes(),2); } },
    // s: Seconds with leading zeros 00-59
    { key: 's', method: function(date) { return enforceLength(date.getSeconds(),2); } }
    // // T: Timezone abbreviation "EST", "MDT", ...
    // { key: 'T', method: function(date) { return date.getTimezone(); } },
    // O: Difference to Greenwich time (GMT) in hours  +0200, -0200
    // { key: 'O', method: function(date) { return date.getGMTOffset(); } },
    // 'P': Difference to Greenwich time (GMT) w/ semicolon between hours:minutes +02:00
    // { key: 'P', method: function(date) { var offset = date.getGMTOffset(); return offset.slice(0,3)+':'+offset.slice(3); } }
  ];

  /**
   * generic formatDate function which accepts dynamic templates
   * @param date {Date} Required
   * @param template {String} Optional
   * @returns {String}
   *
   * pass a date and a template, such as:
   *
   *    util.formatDate(new Date(), '#{M}. #{j}, #{Y}')
   *
   */
  pub.formatDate = function(date,template) {
    template = (typeof template == 'undefined') ? '#{m}/#{d}/#{Y}' : template;
    for(var i=0; i < acceptedDateTokens.length; ++i) {
      if(template.indexOf('#{'+acceptedDateTokens[i].key+'}') > -1) {
        template = pub.injectStringData(template,acceptedDateTokens[i].key,acceptedDateTokens[i].method(date));
      }
    }
    for(var i=0; i < acceptedTimeTokens.length; ++i) {
      if(template.indexOf('#{'+acceptedTimeTokens[i].key+'}') > -1) {
        template = pub.injectStringData(template,acceptedTimeTokens[i].key,acceptedTimeTokens[i].method(date));
      }
    }
    return template;
  };

  return pub;

})((typeof exports === 'undefined') ? {} : exports);


rollup.config.js
import resolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import { uglify } from "rollup-plugin-uglify";
import buble from 'rollup-plugin-buble';
import pkg from './package.json';

let plugins = [
	resolve(),
	commonjs(), 
  buble(),
  uglify()
]; 

export default [
	// browser-friendly UMD build
	{
		input: 'src/index.js',
		output: {
			name: 'timeUtils',
			file: pkg.browser,
			format: 'umd',
			sourcemap: true
		},
		plugins
	},

	// CommonJS (for Node) and ES module (for bundlers) build.
	// (We could have three entries in the configuration array
	// instead of two, but it's quicker to generate multiple
	// builds from a single configuration where possible, using
	// an array for the `output` option, where we can specify 
	// `file` and `format` for each target)
	{
		input: 'src/index.js',
		external: [],
		output: [
			{ file: pkg.main, format: 'cjs' },
			{ file: pkg.module, format: 'es' }
		]
	}
];

timeUtil.js
var timeUtil = (function(w,d,pub){

  /**
   * generic function to inject data into token-laden string
   * @param str {String} Required
   * @param name {String} Required
   * @param value {String|Integer} Required
   * @returns {String}
   *
   * Pass this function a string with encoded tokens, example: 
   *    util.injectStringData("The following is a token: #{tokenName}", "tokenName", 123); 
   *   
   *    Returns: "The following is a token: 123"
   *
   */
  pub.injectStringData = function(str,name,value) {
    return str.replace(new RegExp('#{'+name+'}','g'),value);
  };

  /**
   * generic function to enforce length of string
   * @param str {String} Required
   * @param length {Integer} Required
   * @param fromBack {Boolean} Optional
   * @returns {String}
   *
   * Pass a string or number to this function and specify the desired length.
   * This function will either pad the # with leading 0's (if str.length < length)
   * or remove data from the end (@fromBack==false) or beginning (@fromBack==true)
   * of the string when str.length > length.
   *
   * When length == str.length or typeof length == 'undefined', this function
   * returns the original @str parameter.
   *
   */
  var enforceLength = function(str,length,fromBack) {
    str = str.toString();
    if(typeof length == 'undefined') return str;
    if(str.length == length) return str;
    fromBack = (typeof fromBack == 'undefined') ? false : fromBack;
    if(str.length < length) {
      // pad the beginning of the string w/ enough 0's to reach desired length:
      while(length - str.length > 0) str = '0' + str;
    } else if(str.length > length) {
      if(fromBack) {
        // grab the desired #/chars from end of string: ex: '2015' -> '15'
        str = str.substring(str.length-length);
      } else {
        // grab the desired #/chars from beginning of string: ex: '2015' -> '20'
        str = str.substring(0,length);
      }
    }
    return str;
  };

  // Internal variables for storing days of week, months of year: 
  var daysOfWeek = [ 'Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday' ];
  var monthsOfYear = [ 'January','February','March','April','May','June','July','August','September','October','November','December'];


  var acceptedDateTokens = [
    // d: day of the month, 2 digits with leading zeros:
    { key: 'd', method: function(date) { return enforceLength(date.getDate(), 2); } },
    // D: textual representation of day, 3 letters: Sun thru Sat
    { key: 'D', method: function(date) { return enforceLength(daysOfWeek[date.getDay()],3); } },
    // j: day of month without leading 0's
    { key: 'j', method: function(date) { return date.getDate(); } },
    // l: full textual representation of day of week: Sunday thru Saturday
    { key: 'l', method: function(date) { return daysOfWeek[date.getDay()]; } },
    // F: full text month: 'January' thru 'December'
    { key: 'F', method: function(date) { return monthsOfYear[date.getMonth()]; } },
    // m: 2 digit numeric month: '01' - '12':
    { key: 'm', method: function(date) { return enforceLength(date.getMonth()+1,2); } },
    // M: a short textual representation of the month, 3 letters: 'Jan' - 'Dec'
    { key: 'M', method: function(date) { return enforceLength(monthsOfYear[date.getMonth()],3); } },
    // n: numeric represetation of month w/o leading 0's, '1' - '12':
    { key: 'n', method: function(date) { return date.getMonth() + 1; } },
    // Y: Full numeric year, 4 digits
    { key: 'Y', method: function(date) { return date.getFullYear(); } },
    // y: 2 digit numeric year:
    { key: 'y', method: function(date) { return enforceLength(date.getFullYear(),2,true); } }
  ];

  var acceptedTimeTokens = [
    // a: lowercase ante meridiem and post meridiem 'am' or 'pm'
    { key: 'a', method: function(date) { return (date.getHours() > 11) ? 'pm' : 'am'; } },
    // A: uppercase ante merdiiem and post meridiem 'AM' or 'PM'
    { key: 'A', method: function(date) { return (date.getHours() > 11) ? 'PM' : 'AM'; } },
    // g: 12-hour format of an hour without leading zeros 1-12
    { key: 'g', method: function(date) { return date.getHours%12; } },
    // G: 24-hour format of an hour without leading zeros 0-23
    { key: 'G', method: function(date) { return date.getHours(); } },
    // h: 12-hour format of an hour with leading zeros 01-12
    { key: 'h', method: function(date) { return enforceLength(date.getHours()%12,2); } },
    // H: 24-hour format of an hour with leading zeros: 00-23
    { key: 'H', method: function(date) { return enforceLength(date.getHours(),2); } },
    // i: Minutes with leading zeros 00-59
    { key: 'i', method: function(date) { return enforceLength(date.getMinutes(),2); } },
    // s: Seconds with leading zeros 00-59
    { key: 's', method: function(date) { return enforceLength(date.getSeconds(),2); } },
    // T: Timezone abbreviation "EST", "MDT", ...
    { key: 'T', method: function(date) { return date.getTimezone(); } },
    // O: Difference to Greenwich time (GMT) in hours  +0200, -0200
    { key: 'O', method: function(date) { return date.getGMTOffset(); } }
    // 'P': Difference to Greenwich time (GMT) w/ semicolon between hours:minutes +02:00
    { key: 'P', method: function(date) { var offset = date.getGMTOffset(); return offset.slice(0,3)+':'+offset.slice(3); } }
  ];

  /**
   * generic formatDate function which accepts dynamic templates
   * @param date {Date} Required
   * @param template {String} Optional
   * @returns {String}
   *
   * pass a date and a template, such as:
   *
   *    util.formatDate(new Date(), '#{M}. #{j}, #{Y}')
   *
   */
  pub.formatDate = function(date,template) {
    template = (typeof template == 'undefined') ? '#{m}/#{d}/#{Y}' : template;
    for(var i=0; i < acceptedDateTokens.length; ++i) {
      if(template.indexOf('#{'+acceptedDateTokens[i].key+'}') > -1) {
        template = pub.injectStringData(template,acceptedDateTokens[i].key,acceptedDateTokens[i].method(date));
      }
    }
    return template;
  };

  // end date formatting section

  /**
   * generic popup window fn
   * @param url {String} Required
   * @param title {String} Optional
   * @param sizes {Object} Required
   * @returns {boolean}
   */
  pub.popupWindow = function(url, title, sizes) {
    var width, height, top, left;
    if (typeof url == 'undefined') return false;
    if (typeof title == 'object'
      && typeof sizes == 'undefined') {
      sizes = title;
      title = '';
    }
    if (typeof sizes == 'undefined') {
      sizes = {
        width: 550,
        height: 375
      };
    }
    left = (w.screen.width / 2) - (sizes.width / 2);
    top = (w.screen.height / 2) - (sizes.height / 2);
    w.setTimeout(function(){
      w.open(url, title, 'toolbar=no, location=no, directories=no, status=no, menubar=no, scrollbars=no, resizable=no, copyhistory=no, width=' + sizes.width + ', height=' + sizes.height + ', top=' + top + ', left=' + left);
    },10);
    return false;
  };

  return pub;

})(window,document,{});
